Refactoring

AnKo4 opened the pull request:
Completed refactoring
Needs code review

Commit:
Start refactoring

Commit:
Start adding error handling

Commit:
Add decode error handling

Commit:
Сэйв-коммит, осталось дописать метод отображения ошибки во вью-контроллере

Commit:
Add alert

Commit:
Доделал всплывающее окно с предупреждением об ошибке

Commit:
Add cluster components factory

Commit:
Add some protocols and extensions

Commit:
Refactoring in progress...

Commit:
Поменял структуру папок

Commit:
Folders and files tree did change

ivan-tsaryov commented on GoogleMapsTestApp/ClusterConfigurator/ClusterManagerConfigurator.swift:
```diff
@@ -0,0 +1,38 @@
+            let mapItem = POIItem(position: position, name: item.name ?? "", snippet: item.snippet ?? "", category: item.category)
+            clusterManager.add(mapItem)
+        }
+        clusterManager.cluster()
```
Единственное о чем я подумал. Стоит ли включать вызов `cluster` в метод `configure`. Может быть стоит вызывать отдельно?  Это уже как бы следующая функция, не "конфигурирование"

ivan-tsaryov commented on GoogleMapsTestApp/Networking/NetworkDataProviderProtocol.swift:
```diff
@@ -0,0 +1,40 @@
+            do {
+                return try decodeData(data: response.data, to: structure.self)
+            } catch {
+                throw error
```
Если в `catch` ты опять же "бросаешь" ошибку на уровень "выше", может быть стоит не писать конструкцию `do catch` вовсе?

ivan-tsaryov commented on GoogleMapsTestApp/Scenes/Buttons/ViewModel/ButtonSceneViewModel.swift:
```diff
@@ -8,22 +8,17 @@
+
     func route() {
-        router.goToMapsViewController()
+        router.showGoogleMapsViewController()
```
Советую во ViewModel абстрагироваться от названий View и ViewController. Использовать `Module` или просто `showGoogleMaps`

borisoft commented on GoogleMapsTestApp/Enums/ClusterAlgorithm.swift:
```diff
@@ -0,0 +1,25 @@
+
+import Foundation
+
+enum ClusterAlgorithm: String, Codable {
```
`ClusterAlgorithmType` тут скорее,  раз ты хочешь получать его из сервера или хранить в виде json

borisoft replied:
Тут просто изначально можно подумать что получаешь именно алгоритм модель,  которая будет выполняться,  в твоем же случае ты получаешь скорее тип алгоритма который  можешь передать в фабричный метод

borisoft commented on GoogleMapsTestApp/Enums/ClusterAlgorithm.swift:
```diff
@@ -0,0 +1,25 @@
+}
+
+extension ClusterAlgorithm {
+    var value: GMUClusterAlgorithm {
```
и тогда тут не Value а что-то типо algorithmForType или еще как-то

borisoft commented on GoogleMapsTestApp/ClusterConfigurator/GMUComponentFactory.swift:
```diff
@@ -0,0 +1,25 @@
+
+class GMUComponentFactory: GMUComponentFactoryProtocol {
+    
+    func makeIconGenerator(buckets: [NSNumber]?, colors: [UIColor]?) -> GMUClusterIconGenerator {
```
Тут же мы говорили о структуре параметров,  или здесь не все изменения ?

Commit:
Additional changes in folders and files tree
Add protocol and data structure for Icon generator parameters
Add factory method for Icon generator parameters data structure
